### ISSUE TITLE
Changing the animationplayer slider is not entered in the UndoRedo stack

### ISSUE THREAD

**Issue by securas** [godotengine/godot#46123]
**Godot version:**
3.2.4 RC2

**OS/device including version:**
Windows

**Issue description:**
Changing the time slider of an animation is not considered a scene change. Probably it is not entered in the UndoRedo stack.
See proposal godotengine/godot-proposals#2301

**Steps to reproduce:**
Create an animation and a new track with a moving object (for example). Move the time slider. The scene is not marked as changed.

**Minimal reproduction project:**
[test_animation_player.zip](https://github.com/godotengine/godot-proposals/files/5992858/test_animation_player.zip)
(from proposal godotengine/godot-proposals#2301)

**Comment by akien-mga** [comment]
Had a quick look at it, here's one possible option to fix the issue by properly registering an undo/redo when seeking the timeline and/or the timestamp spinbox (called `frame` in the code as it can also be used in frames mode).

```diff
diff --git a/editor/plugins/animation_player_editor_plugin.cpp b/editor/plugins/animation_player_editor_plugin.cpp
index a2e7f7640a..dd4f982071 100644
--- a/editor/plugins/animation_player_editor_plugin.cpp
+++ b/editor/plugins/animation_player_editor_plugin.cpp
@@ -1040,13 +1040,16 @@ void AnimationPlayerEditor::_seek_value_changed(float p_value, bool p_set) {
                pos = Math::stepify(pos, _get_editor_step());
        }
 
+       float prev_pos = player->get_current_animation_position();
        if (player->is_valid() && !p_set) {
-               float cpos = player->get_current_animation_position();
-
-               player->seek_delta(pos, pos - cpos);
+               player->seek_delta(pos, pos - prev_pos);
        } else {
-               player->stop(true);
-               player->seek(pos, true);
+               undo_redo->create_action(TTR("Seek Current Animation"));
+               undo_redo->add_do_method(player, "stop", true);
+               undo_redo->add_undo_method(player, "stop", true);
+               undo_redo->add_do_method(player, "seek", pos, true);
+               undo_redo->add_undo_method(player, "seek", prev_pos, true);
+               undo_redo->commit_action();
        }
 
        track_editor->set_anim_pos(pos);
```

It works but it also exposes another issue which warrants digging a bit further (at least in the `3.2` branch which I toyed with): in `AnimationPlayerEditor::_animation_key_editor_seek` (which is the callback for seeking the timeline), the `_seek_value_changed` method is called twice and this registers two UndoRedo actions:
https://github.com/godotengine/godot/blob/f91c4c08999cc79586d74973953202a2af4dd7ee/editor/plugins/animation_player_editor_plugin.cpp#L1070-L1073

The first call is due to `frame->set_value()`, as its `value_changed` signal is also connected to `_seek_value_changed`: https://github.com/godotengine/godot/blob/f91c4c08999cc79586d74973953202a2af4dd7ee/editor/plugins/animation_player_editor_plugin.cpp#L1694

I don't plan to work further on this in the near future but this can serve as a base for anyone interested to dig further.

**Comment by angad-k** [comment]
Will this work? I just registered the undo action for set_value directly.
```diff
diff --git a/editor/plugins/animation_player_editor_plugin.cpp b/editor/plugins/animation_player_editor_plugin.cpp      
index cda88c00f3..3366b840b9 100644                                                                                     
--- a/editor/plugins/animation_player_editor_plugin.cpp                                                                 
+++ b/editor/plugins/animation_player_editor_plugin.cpp                                                                 
@@ -1067,8 +1067,12 @@ void AnimationPlayerEditor::_animation_key_editor_seek(float p_pos, bool p_drag)
                return;
         }                                                                                                                                                                                                                                       
+       float prev_pos = player->get_current_animation_position();                                                      
+       undo_redo->create_action(TTR("Seek Current Animation"));                                                        
+       undo_redo->add_do_method(frame, "set_value", Math::snapped(p_pos, _get_editor_step()));                         
+       undo_redo->add_undo_method(frame, "set_value", Math::snapped(prev_pos, _get_editor_step()));                            
	updating = true;
-       frame->set_value(Math::snapped(p_pos, _get_editor_step()));                                                     
+       undo_redo->commit_action();                                                                                             
	updating = false;                                                                                                       
	_seek_value_changed(p_pos, !p_drag);                                                                             
} 
```
If yes, then I'll send a PR.

**Comment by akien-mga** [comment]
That will work only for the timeline slider, and not when changing the value with the `frame` SpinBox.

**Comment by akien-mga** [comment]
>  Also, why is ` _seek_value_changed(p_pos, !p_drag);` called again explicitly if `frame->set_value()` does lead to it getting called?

That's what needs to be investigated. Might be unnecessary, or maybe not. The `p_drag` argument is an important part of that puzzle.

**Comment by angad-k** [comment]
Removing the explicit call led to the seek action not leading to anything updating. So, I guess the call is necessary. 
Which method does the spinbox call? What if we add [this](https://github.com/godotengine/godot/issues/46123#issuecomment-781310340) and another undo-redo action in the method which the spin box calls?

**Comment by angad-k** [comment]
Another interesting thing I noticed now. I added warning statements to check the flow of code in the if and else [here](https://github.com/godotengine/godot/issues/46123#issuecomment-780645071) and found that two warnings are sent out when I comment out `frame->set_value(Math::snapped(p_pos, _get_editor_step()))` and no warnings are sent out if the explicit call `_seek_value_changed(p_pos, !p_drag)` instead.

**Comment by angad-k** [comment]
How about this : 
```diff
diff --git a/editor/plugins/animation_player_editor_plugin.cpp b/editor/plugins/animation_player_editor_plugin.cpp      
index cda88c00f3..eaf9f75968 100644                                                                                     
--- a/editor/plugins/animation_player_editor_plugin.cpp                                                                 
+++ b/editor/plugins/animation_player_editor_plugin.cpp                                                                 
@@ -1015,13 +1015,19 @@ void AnimationPlayerEditor::_seek_value_changed(float p_value, bool p_set) {
                                    pos = Math::snapped(pos, _get_editor_step());
        }
+       float prev_pos = player->get_current_animation_position();
        if (player->is_valid() && !p_set) {                                                                             
-               float cpos = player->get_current_animation_position();                                                  
-                                                                                                                       
-               player->seek_delta(pos, pos - cpos);                                                                    
+               player->seek_delta(pos, pos - prev_pos);                                                                        
	} else {                                                                                                        
-               player->stop(true);   
-               player->seek(pos, true);                                                                                
+               if(pos != prev_pos)                                                                                     
+               {                                                                                                       
+                       undo_redo->create_action(TTR("Seek Current Animation"));                                        
+                       undo_redo->add_do_method(player, "stop", true);                                                 
+                       undo_redo->add_undo_method(player, "stop", true);                                               
+                       undo_redo->add_do_method(player, "seek", pos, true);                                            
+                       undo_redo->add_undo_method(player, "seek", prev_pos, true);                                     
+                       undo_redo->commit_action();              
+               }                                                                                                               
	}                                                                                                                                                                                                                                               
	track_editor->set_anim_pos(pos);   
```
Tried moving the animation slider and also changed the frame through the spin box. Works for both of them.

**Comment by angad-k** [comment]
I have made the changes in the PR so that the undo-redo action is registered only once and everything properly except that the value in spin box is not changing after undo. Any idea why?

**Comment by angad-k** [comment]
Done with the PR. Undo now works flawlessly for animation player seek and also on changing the value in spin box. Someone please review...

**Comment by KoBeWi** [comment]
This is not a bug, and it was somewhat resolved by the addition of RESET animations. And as I mentioned in the linked PR, playing the animation would also need to be registered as undo action, which comes with some problems.

If you are still interested having this issue "fixed", open a proposal instead.
